madara/crates/client/settlement_client/src/eth/mod.rs

```
@@ -49,30 +52,92 @@ pub struct EthereumClientConfig {
+        EthereumClient {
+            provider: Arc::clone(&self.provider),
+            l1_core_contract: self.l1_core_contract.clone(),
+            health: Arc::clone(&self.health),
+        }
     }
 }
 
 impl EthereumClient {
+    /// Create a new Ethereum client with lazy initialization.
+    ///
+    /// This method no longer blocks on L1 availability. Instead, it:
+    /// - Validates the contract address format (fails fast on config errors)
+    /// - Creates the provider and contract instances
+    /// - Defers actual L1 connection verification to the first RPC call
+    ///
+    /// This allows Madara to start even when L1 is temporarily unavailable.
+    /// The retry logic in individual RPC calls will handle connection failures.
     pub async fn new(config: EthereumClientConfig) -> Result<Self, SettlementClientError> {
         let provider = ProviderBuilder::new().on_http(config.rpc_url);
         let core_contract_address =
             Address::from_str(&config.core_contract_address).map_err(|e| -> SettlementClientError {
                 EthereumClientError::Conversion(format!("Invalid core contract address: {e}")).into()
             })?;
-        // Check if contract exists
-        if !provider
-            .get_code_at(core_contract_address)
-            .await
-            .map_err(|e| -> SettlementClientError { EthereumClientError::Rpc(e.to_string()).into() })?
-            .is_empty()
-        {
-            let contract = StarknetCoreContract::new(core_contract_address, provider.clone());
-            Ok(Self { provider: Arc::new(provider), l1_core_contract: contract })
-        } else {
-            Err(SettlementClientError::Ethereum(EthereumClientError::Contract(
-                "Core contract not found at given address".into(),
-            )))
+
+        // Note: We no longer check if the contract exists here to avoid blocking startup
+        // The contract existence will be verified on the first RPC call, with retry logic
```

so if we have wrong contract it will keep retrying? I don't think that makes sense, atleast for startup we can remove this logic of retry IMO and it makes sense as well, given this new function would be called in the begining only

---

Fixed! Added contract verification at startup using get_code_at(). If the contract doesn't exist, it fails fast with a clear error message instead of retrying indefinitely.